text_quality/language/fasttext.py

```diff
@@ -0,0 +1,120 @@
+            self._download_model(model_file)
+        self._model = fasttext.load_model(str(model_file))
+
+        self._line_threshold = line_threshold
+
+    def classify(self, text: str) -> tuple[str, float]:
+        """Classify a text string.
+
+        Args:
+            text (str): The text to classify.
+        Returns:
+            A tuple[str, float] with the language code (e.g. "nl") and the confidence.
+        """
+
+        lines: list[str] = [
+            LanguageClassifier.preprocess(line) for line in text.split("\n")
+        ]
+        line_labels, line_confidences = self._model.predict(
+            lines, threshold=self._line_threshold
+        )
+        language, confidence = FastTextLanguageClassifier._aggregate_lines(
+            line_labels, line_confidences
+        )
+
+        return language.removeprefix(self._LABEL_PREFIX), confidence
+
+    @staticmethod
+    def _download_model(model_file: Path):
+        try:
+            url = FastTextLanguageClassifier.MODEL_URLS[model_file.name]
+        except KeyError as e:
+            raise ValueError(
+                f"No download location known for '{model_file.name}'."
+            ) from e
+
+        model_file.parent.mkdir(parents=True, exist_ok=True)
+
+        logging.info("Downloading from '%s', storing at '%s' .", model_file, url)
+        with urllib.request.urlopen(url) as response, model_file.open("wb") as out_file:
+            shutil.copyfileobj(response, out_file)
+
+    @staticmethod
+    def _aggregate_lines(
+        line_labels: list[list[str]], line_confidences: list[ArrayLike]
+    ) -> tuple[str, float]:
+        """Aggregate the results per line from the classifier.
+
+        Args:
+            line_labels (list[list[str]]): the labels returned by the classifier; one list of labels for each input line
+            line_confidences (list[ArrayLike]): the confidences returned by the classifier; one array for each input line
+        Returns:
+            A tuple[str, float] with the language code (e.g. "nl") and the confidence.
+                ("None", 0.0) if no input line could be classified.
+        """
+        if len(line_labels) != len(line_confidences):
+            raise ValueError(
+                f"Labels and confidences must have the same length, but were {len(line_labels)} and {len(line_confidences)}."
+            )
+
+        if any(labels for labels in line_labels):
+            total_confidences = Counter()
+            for labels, confidences in zip(line_labels, line_confidences):
+                # Iterate over labels and confidences per input line
+                # Number of entries per line corresponds to k parameter in model.predict(), defaults to 1
+                for label, confidence in zip(labels, confidences):
+                    total_confidences[label] += confidence
+            label, total_confidence = total_confidences.most_common()[0]
+            confidence = total_confidence / sum(total_confidences.values())
```

Comment:
This calculation seems off to me. Can you explain your reasoning?

I played with it a bit; may be useful to illustrate why I'm confused:
Say, there are three lines, classified as `(language, confidence)`:
```
("nl", 0.8),
("en", 0.1),
("nl", 0.7)
```

With the calculation you do, as far as I understand, this would get to the resulting confidence as
```
(0.8 + 0.7) / (0.8 + 0.1 + 0.7) = 1.5 / 1.6 = 0.9375
```

Or: if there are low confidence levels for all lines resulting from the original classifier, but all the same language, the resulting confidence is 1.

Is that intended?

Reply:
Your calculation is correct. The intuition behind it is to get the weight that the classifier has had per line. So indeed, if all lines have the same winner, the confidence is always 1.

This has been motivated by the observation, that many obscure lines (e.g. almost exclusively punctuation) are all assigned to some arbitrary language, but with low confidence. I try to give these lines with arbitrary "winners" little weight.

Anyway, the confidence is barely used at this point. We need to run some experiments to get empirical evidence, and adapt this calculation accordingly.

Reply:
Got it!
Maybe add a few comment lines to explain this reasoning, or link to this review?

I do also wonder: what if there are very few lines in the text; the confidence could be 1 if all of them by chance are classified as the same language, but this would not match the actual confidence. Perhaps there should be a punishment for a low number of lines (there is a cut-off for minimum text length already, but I don't know how much that is), and still some inclusion of the average confidence for the most frequently used language...

That said, this may complicate things too much, and the idea is really to have a quick check of the most likely language, correct?

Reply:
Correct, I think the confidence will only play a role in edge cases anyway, so it is very hard to predict what would lead to best results without empirical evaluation.
Also, there is a threshold set in the `classify()` (to be renamed) method above, that excludes all predictions with a confidence below 0.5 (default value). I will add more documentation about the intuition, and suggest leaving it as it is for now, pending on user feedback/evaluation.